ghost/email-service/lib/mailgun-email-provider.js

```diff
@@ -0,0 +1,168 @@
+
+    static BATCH_SIZE = 1000;
+
+    /**
+     * @param {object} dependencies
+     * @param {import('@tryghost/mailgun-client/lib/mailgun-client')} dependencies.mailgunClient - mailgun client to send emails
+     * @param {Function} [dependencies.errorHandler] - custom error handler for logging exceptions
+     */
+    constructor({
+        mailgunClient,
+        errorHandler
+    }) {
+        this.#mailgunClient = mailgunClient;
+        this.#errorHandler = errorHandler;
+    }
+
+    #createRecipientData(replacements) {
+        let recipientData = {};
+
+        recipientData = replacements.reduce((acc, replacement) => {
+            const {id, value} = replacement;
+            if (!acc[id]) {
+                acc[id] = {};
+            }
+            acc[id] = value;
+            return acc;
+        }, {});
+
+        return recipientData;
+    }
+
+    #updateRecipientVariables(data, replacementDefinitions) {
+        for (const def of replacementDefinitions) {
+            data = data.replace(
+                def.token,
+                `%recipient.${def.id}%`
+            );
+        }
+        return data;
+    }
+
+    /**
+     * Create mailgun error message for storing in the database
+     * @param {Object} error
+     * @param {string} error.message
+     * @param {string} error.details
+     * @returns {string}
+    */
+    #createMailgunErrorMessage(error) {
+        const message = (error?.message || '') + ':' + (error?.details || '');
+        return message.slice(0, 2000);
+    }
+
+    /**
+     * Send an email using the Mailgun API
+     * @param {import('./sending-service').EmailData} data
+     * @param {EmailSendingOptions} options
+     * @returns {Promise<EmailProviderSuccessResponse>}
+    */
+    async send(data, options) {
+        const {
+            subject,
+            html,
+            plaintext,
+            from,
+            replyTo,
+            id,
+            recipients,
+            replacementDefinitions
+        } = data;
+
+        logging.info(`Sending email to ${recipients.length} recipients`);
+        const startTime = Date.now();
+        debug(`sending message to ${recipients.length} recipients`);
+
+        try {
+            const messageData = {
+                subject,
+                html,
+                plaintext,
+                from,
+                replyTo,
+                id,
+                track_opens: !!options.openTrackingEnabled,
+                track_clicks: !!options.clickTrackingEnabled
+            };
+
+            // create recipient data for Mailgun using replacement definitions
+            const recipientData = recipients.reduce((acc, recipient) => {
+                acc[recipient.email] = this.#createRecipientData(recipient.replacements);
+                return acc;
+            }, {});
+
+            // update content to use Mailgun variable syntax for all replacements
+            ['html', 'plaintext', 'subject'].forEach((key) => {
```

Comment:
Cool, so we could use 'smart' variables inside subjects too 😊 Would need an update in the email renderer, but otherwise this could be supported

Reply:
So I was confused about `subject` originally, but I saw we allow subject to be transformed [here](https://github.com/TryGhost/Ghost/blob/9fe1274fbe23ec40cb88dc9b5a01e5415fee3f4b/ghost/mailgun-client/lib/mailgun-client.js#L49) which is why I kept it here. Tho we don't add the `subject` to formats [here](https://github.com/TryGhost/Ghost/blob/65d910f43b3e2962fffe3acbd0d8c964401ffbbb/ghost/core/core/server/services/mega/post-email-serializer.js#L191) so it isn't getting used right now, but i think it can potentially be supported.

Reply:
Ah, yes, but in the past the format was never set to subject

Reply:
Yea exactly, should we remove the `subject` from this list for now then ? Don't want accidentaly coverting a {{name}} in subject to the member name i guess right